### PR TITLE
HAWQ-462. Dispatch dfs_address from master to segment in pxf secure f…

### PR DESCRIPTION
**Pull request by kavinderd**
…ilesystem case

Set global var dfs_address on segment after deserialization

### CODE REVIEW

**Comment by kavinderd** [comment]
@hornn @shivzone  Alternate implementation that handles both readable and writable case by setting a global var on segment after deserialization

I declared the var `dfs_address` in `fd.h` since that is where `enable_filesystem_credential` is declared there as well

File: src/bin/gpfusion/gpbridgeapi.c [hunk 1]
```diff
@@ -508,15 +508,12 @@ size_t fill_buffer(gphadoop_context* context, char* start, size_t size)
 void add_delegation_token(PxfInputData *inputData)
 {
 	PxfHdfsTokenData *token = NULL;
-	char* dfs_address = NULL;
 
 	if (!enable_secure_filesystem)
 		return;
```
Comment (hornn) [inline src/bin/gpfusion/gpbridgeapi.c]: perhaps add assert that dfs_address is not null

File: src/backend/cdb/cdbquerycontextdispatching.c [hunk 2]
```diff
@@ -2984,3 +3013,20 @@ GetResultRelSegFileInfos(Oid relid, List *segnomaps, List *existing_seginfomaps)
 
 	return existing_seginfomaps;
 }
+
+void
+prepareDfsAddressForDispatch(QueryContextInfo* cxt){
+//	if (enable_secure_filesystem)
```
Comment (hornn) [inline src/backend/cdb/cdbquerycontextdispatching.c]: remove comments?

File: src/backend/cdb/cdbquerycontextdispatching.c [hunk 3]
```diff
@@ -770,6 +770,31 @@ RebuildTupleForRelation(QueryContextInfo *cxt)
 }
 
 /*
+ * Deserialize the Namespace data
+ */
+
```
Comment (hornn) [inline src/backend/cdb/cdbquerycontextdispatching.c]: remove new line

**Comment by hornn** [comment]
I really like this solution - it looks really clean. 
What happens in the non secured case - do we always pass this var regardless?

File: src/backend/cdb/cdbquerycontextdispatching.c (hunk 2 above)
Comment (kavinderd) [inline src/backend/cdb/cdbquerycontextdispatching.c]: oh yeah, oops

**Comment by kavinderd** [comment]
In the non-secure case the `dfs_address` is not serialized and dispatched from the master so on the segment the global var will just be `NULL`

File: src/backend/cdb/cdbquerycontextdispatching.c [hunk 4]
```diff
@@ -2984,3 +3012,21 @@ GetResultRelSegFileInfos(Oid relid, List *segnomaps, List *existing_seginfomaps)
 
 	return existing_seginfomaps;
 }
+
+void
+prepareDfsAddressForDispatch(QueryContextInfo* cxt)
+{
+	if (enable_secure_filesystem)
```
Comment (shivzone) [inline src/backend/cdb/cdbquerycontextdispatching.c]: This should be ! enable_secure_filesystem

File: src/backend/cdb/cdbquerycontextdispatching.c [hunk 5]
```diff
@@ -2984,3 +3012,21 @@ GetResultRelSegFileInfos(Oid relid, List *segnomaps, List *existing_seginfomaps)
 
 	return existing_seginfomaps;
 }
+
+void
+prepareDfsAddressForDispatch(QueryContextInfo* cxt)
```
Comment (shivzone) [inline src/backend/cdb/cdbquerycontextdispatching.c]: add comment to describe the purpose of this function

File: src/backend/cdb/cdbquerycontextdispatching.c [hunk 6]
```diff
@@ -801,6 +825,9 @@ RebuildQueryContext(QueryContextInfo *cxt, HTAB **currentFilesystemCredentials,
         	RebuildFilesystemCredentials(cxt, currentFilesystemCredentials,
         	        currentFilesystemCredentialsMemoryContext);
         	break;
+        case Namespace:
+            RebuildNamespace(cxt);
```
Comment (shivzone) [inline src/backend/cdb/cdbquerycontextdispatching.c]: Will this case be invoked in the non secure case where you don't set the namespace

**Comment by shivzone** [comment]
+1

File: src/backend/cdb/cdbquerycontextdispatching.c (hunk 6 above)
Comment (kavinderd) [inline src/backend/cdb/cdbquerycontextdispatching.c]: No, it won't. The `Namespace` enum is only added to the buffer in the case of a secure filesystem

File: src/backend/cdb/cdbquerycontextdispatching.c [hunk 7]
```diff
@@ -1747,6 +1774,7 @@ prepareDispatchedCatalogExternalTable(QueryContextInfo *cxt,
 	{
 		Insist(array_size == 1);
 		AddFileSystemCredentialForPxfTable(location);
```
Comment (GodenYao) [inline src/backend/cdb/cdbquerycontextdispatching.c]: is this still needed? AddFileSystemCredentialForPxfTable as you dispatch it through query context now.

**Comment by GodenYao** [comment]
+1

**Comment by kavinderd** [comment]
Merged into master by @shivzone [here](https://github.com/apache/incubator-hawq/commit/59ebfa7072621117827ae3d9464c971a61919672)

**Comment by ictmalili** [comment]
@shivzone @kavinderd , actually this commit blocks build on MAC. Since the function is referenced before the declaration.
{code}
cdbquerycontextdispatching.c:3026:1: error: conflicting types for 'prepareDfsAddressForDispatch'
prepareDfsAddressForDispatch(QueryContextInfo* cxt)
^
cdbquerycontextdispatching.c:1777:3: note: previous implicit declaration is here
                prepareDfsAddressForDispatch(cxt);
                ^
2 warnings and 1 error generated.
make[3]: *** [cdbquerycontextdispatching.o] Error 1
make[2]: *** [cdb-recursive] Error 2
make[1]: *** [all] Error 2
make: *** [all] Error 2
{code}

**Comment by kavinderd** [comment]
@ictmalili will fix asap

**Comment by ictmalili** [comment]
@kavinderd  I have fixed this. Could you see the pull request? https://github.com/apache/incubator-hawq/pull/554

**Comment by kavinderd** [comment]
@ictmalili Oh ok cool. Thanks for taking care of it

File: src/backend/cdb/cdbquerycontextdispatching.c [hunk 8]
```diff
@@ -770,6 +770,30 @@ RebuildTupleForRelation(QueryContextInfo *cxt)
...
+	char buffer[4], *binary;
+	ReadData(cxt, buffer, sizeof(buffer), TRUE);
+
+	len = (int) ntohl(*(uint32 *) buffer);
+	binary = palloc(len);
+	if(ReadData(cxt, binary, len, TRUE))
+	{
+		StringInfoData buffer;
+		initStringInfoOfString(&buffer, binary, len);
+		dfs_address = strdup(buffer.data);
```
Comment (ztao1987) [inline src/backend/cdb/cdbquerycontextdispatching.c]: Why not just strdup(binary) to dfs_address?

Comment (kavinderd) [inline src/backend/cdb/cdbquerycontextdispatching.c]: Good point, no need for `buffer`. I'll update this

File: src/bin/gpfusion/gpbridgeapi.c [hunk 9]
```diff
@@ -536,3 +536,8 @@ void free_token_resources(PxfInputData *inputData)
 
 	pfree(inputData->token);
 }
+
+void free_dfs_address()
+{
+	free(dfs_address);
+}
```
Comment (ztao1987) [inline src/bin/gpfusion/gpbridgeapi.c]: Will free_dfs_address only be called in secure mode? Otherwise it will coredump

Comment (kavinderd) [inline src/bin/gpfusion/gpbridgeapi.c]: Yes only in the secure mode. However it is only dispatched from master in the secure mode as well so it is never allocated in any other case

Comment (ztao1987) [inline src/bin/gpfusion/gpbridgeapi.c]: I checked free_dfs_address is called by gpbridge_export_start and gpbridge_import_start, do you mean this two funtions are only called in the secure mode? If not, you are freeing NULL pointer.

Comment (kavinderd) [inline src/bin/gpfusion/gpbridgeapi.c]: Oh right, good catch. I will add a check in `free_dfs_address` for secure filesystem. Sound good?

Comment (ztao1987) [inline src/bin/gpfusion/gpbridgeapi.c]: Yes, just check NULL condition is enough.

Comment (kavinderd) [inline src/bin/gpfusion/gpbridgeapi.c]: Sure, thanks

**Comment by ictmalili** [comment]
Adding @ztao1987 's comments, LGTM.

**Pull request closed without merging**